courses/fundamentals_of_ada/adv_270_subprogram_contracts.rst

```diff
@@ -35,227 +35,262 @@ Subprogram Contracts
 Introduction
 ==============
 
-----------------------------
-Contract-Based Programming
-----------------------------
+--------------------------
+:dfn:`Design-By-Contract`
+--------------------------
 
 * Source code acting in roles of **client** and **supplier** under a binding **contract**
 
-   - *Supplier* provides services
-   - *Client* utilitizes services
-   - *Contract* specifies requirements and guarantees
+   - :dfn:`Contract` specifies *requirements* or *guarantees*
+
+      - *"A specification of a software element that affects its use by potential clients."* (Bertrand Meyer)
+
+   - :dfn:`Supplier` provides services
+
+       - In Ada: Implements it
+       - Guarantees specific functional behavior
+       - Has requirements for guarantees to hold
+
+   - :dfn:`Client` utilizes services
+
+       - In Ada: Calls the subprogram
+       - Guarantees supplier's conditions are met
+       - Requires result to follow the subprogram's guarantees
 
-      - "A specification of a software element that affects its use by potential clients." (Bertrand Meyer)
+---------------
+Ada Contracts
+---------------
 
-* Includes enforcement
+* Ada contracts include enforcement
 
    - At compile-time: specific constructs, features, and rules
    - At run-time: language-defined and user-defined exceptions
 
-------------------
-Contracts In Ada
-------------------
+* Facilities prior to Ada 2012
 
-* Exist implicitly in facilities you may already be using
-
-   - Exceptions
    - Range specifications
-   - Subtypes
    - Parameter modes
-   - OOP interface types
-   - et cetera
+   - Generic contracts
+   - OOP :ada:`interface` types (Ada 2005)
+   - Work well, but on a restricted set of use-cases
 
-* Are explicitly supported
+* Contracts are explicitly added in **Ada 2012**
```

Comment:
Not true.
Contract _aspects_ added in 2012. "pragma pre/post" have been around since the beginning (hence the fact that SPARK has been around for almost as long as Ada)